File: EEntityCore.DB/EEntityCore.VB.MSSQL.WebTest/DBEntities/DatabaseSchema/DataColumnDefinition.vb


Namespace DBEntities.DatabaseSchema

    Public Class DataColumnDefinition

#Region "Constructors"
        Sub New(ByVal pColumnName As String,
                ByVal pDataType As System.Type,
                ByVal pNullable As Boolean,
                ByVal pDefaultValue As Object,
                Optional ByVal pConstraintType As ConstraintTypes = ConstraintTypes.UNKNOWN)

            Me.vNullable = pNullable
            Me.vDefaultValue = pDefaultValue
            Me.vDataType = pDataType
            Me.vColumnName = pColumnName
            Me.vConstraintType = pConstraintType

        End Sub

#End Region


#Region "Constructors"

        Private vColumnName As String, vDataType As System.Type, vDefaultValue As Object, vNullable As Boolean, vConstraintType As ConstraintTypes

        Public ReadOnly Property ColumnName As String
            Get
                Return Me.vColumnName
            End Get
        End Property
        Public ReadOnly Property DataType As System.Type
            Get
                Return Me.vDataType
            End Get
        End Property


        Public ReadOnly Property DefaultValue As Object
            Get
                If Not Me.Nullable Then
                    REM WHoop up something
                    Select Case DataColumnDefinition.getTypeAllowed(Me.DataType)

                        Case DataColumnDefinition.AllowedDataTypes.Bool
                            Return False

                        Case DataColumnDefinition.AllowedDataTypes.Blob
                            Return New Byte() {}

                        Case DataColumnDefinition.AllowedDataTypes.DateTime
                            Return DateTime.Now

                        Case DataColumnDefinition.AllowedDataTypes.Decimal, DataColumnDefinition.AllowedDataTypes.Int, DataColumnDefinition.AllowedDataTypes.Long
                            Return 0

                        Case DataColumnDefinition.AllowedDataTypes.String
                            Return " "              REM Some database sees "" as NULL

                        Case DataColumnDefinition.AllowedDataTypes.TimeSpan
                            Throw New NotImplementedException("Time Span is not implemented")

                        Case Else
                            REM DataColumnDefinition.AllowedDataTypes.UNKNOWN()
                            Throw New NotImplementedException("UNKNOWN DATA TYPE")

                    End Select

                Else
                    REM If it is date and is not null or nothing return now
                    ''If DataColumnDefinition.getTypeAllowed(Me.DataType) = AllowedDataTypes.DateTime AndAlso
                    ''    (Me.vDefaultValue IsNot Nothing OrElse Not IsDBNull(Me.vDefaultValue)) Then Return DateTime.Now


                    Return Me.vDefaultValue
                End If
            End Get
        End Property

        Public ReadOnly Property Nullable As Boolean
            Get
                Return Me.vNullable
            End Get
        End Property

        Public ReadOnly Property ConstraintType As ConstraintTypes
            Get
                Return Me.vConstraintType
            End Get
        End Property

#End Region


#Region "Enums and Consts"


        Public Enum AllowedDataTypes
            [Bool]
            [Int]
            [Long]
            [Decimal]
            [String]
            [Blob]
            [DateTime]
            [TimeSpan]
            [UNKNOWN]
        End Enum

        Public Enum ConstraintTypes
            FOREIGN
            PRIMARY
            UNIQUE
            CHECK
            UNKNOWN
        End Enum

#End Region


        Public Shared Function getTypeAllowed(pType As Type) As AllowedDataTypes
            Select Case pType
                Case GetType(Boolean)
                    Return AllowedDataTypes.Bool

                Case GetType(Byte), GetType(SByte), GetType(Int16), GetType(UInt16), GetType(Int32), GetType(UInt32)
                    Return AllowedDataTypes.Int

                Case GetType(Int64), GetType(UInt64)
                    Return AllowedDataTypes.Long

                Case GetType(Decimal), GetType(Double)
                    Return AllowedDataTypes.Decimal

                Case GetType(String), GetType(Char)
                    Return AllowedDataTypes.String

                Case GetType(Byte())
                    Return AllowedDataTypes.Blob

                Case GetType(DateTime)
                    Return AllowedDataTypes.DateTime

                Case GetType(TimeSpan)
                    Return AllowedDataTypes.TimeSpan

                Case Else
                    Return AllowedDataTypes.UNKNOWN

            End Select
        End Function



    End Class



End Namespace

File: EEntityCore.DB/EEntityCore.DB/Schemas/SQLServerSchema/DataColumnDefinition.cs
using System;

namespace EEntityCore.DB.Schemas.SQLServerSchema
{
    public class DataColumnDefinition
    {

        #region Constructors
        public DataColumnDefinition(
            IDatabaseInit pointerDatabaseInit,
            string pColumnName, Type pDataType, 
            bool pNullable, object pDefaultValue,
            ConstraintTypes pConstraintType = ConstraintTypes.UNKNOWN
            )
        {
            this.PointerDatabaseInit = pointerDatabaseInit;
            vNullable = pNullable;
            vDefaultValue = pDefaultValue;
            vDataType = pDataType;
            vColumnName = pColumnName;
            vConstraintType = pConstraintType;
        }

        #endregion


        #region Properties

        // This part should adjust after it has been moved to MSSQL
        private IDatabaseInit PointerDatabaseInit { get; }
        public IDatabaseInit DatabaseInit { get { return PointerDatabaseInit.GetDatabaseInit(); } }

        private string vColumnName;
        private Type vDataType;
        private object vDefaultValue;


        private bool vNullable;
        private ConstraintTypes vConstraintType;

        public string ColumnName
        {
            get
            {
                return vColumnName;
            }
        }

        public Type DataType
        {
            get
            {
                return vDataType;
            }
        }

        public object DefaultValue
        {
            get
            {
                if (!Nullable)
                {
                    // REM WHoop up something
                    switch (GetTypeAllowed(DataType))
                    {
                        case AllowedDataTypes.Bool:
                            {
                                return false;
                            }

                        case AllowedDataTypes.Blob:
                            {
                                return new byte[] { };
                            }

                        case AllowedDataTypes.DateTime:
                            {
                                return DateTime.Now;
                            }

                        case AllowedDataTypes.Decimal:
                        case AllowedDataTypes.Int:
                        case AllowedDataTypes.Long:
                            {
                                return 0;
                            }

                        case AllowedDataTypes.String:
                            {
                                return " ";              // REM Some database sees "" as NULL
                            }

                        case AllowedDataTypes.TimeSpan:
                            {
                                throw new NotImplementedException("Time Span is not implemented");
                            }

                        default:
                            {
                                // REM DataColumnDefinition.AllowedDataTypes.UNKNOWN()
                                throw new NotImplementedException("UNKNOWN DATA TYPE");
                            }
                    }
                }
                else
                {
                    // REM If it is date and is not null or nothing return now
                    // 'If DataColumnDefinition.getTypeAllowed(Me.DataType) = AllowedDataTypes.DateTime AndAlso
                    // '    (Me.vDefaultValue IsNot Nothing OrElse Not IsDBNull(Me.vDefaultValue)) Then Return DateTime.Now


                    return vDefaultValue;
                }
            }
        }

        public bool Nullable
        {
            get
            {
                return vNullable;
            }
        }

        public ConstraintTypes ConstraintType
        {
            get
            {
                return vConstraintType;
            }
        }

        #endregion


        #region Enums and Consts


        public enum AllowedDataTypes
        {
            Bool,
            Int,
            Long,
            Decimal,
            String,
            Blob,
            DateTime,
            TimeSpan,
            UNKNOWN
        }

        public enum ConstraintTypes
        {
            FOREIGN,
            PRIMARY,
            UNIQUE,
            CHECK,
            UNKNOWN
        }

        #endregion


        public static AllowedDataTypes GetTypeAllowed(Type pType)
        {
            switch (pType)
            {
                case var @case when @case == typeof(bool):
                    {
                        return AllowedDataTypes.Bool;
                    }

                case var case1 when case1 == typeof(byte):
                case var case2 when case2 == typeof(sbyte):
                case var case3 when case3 == typeof(short):
                case var case4 when case4 == typeof(ushort):
                case var case5 when case5 == typeof(int):
                case var case6 when case6 == typeof(uint):
                    {
                        return AllowedDataTypes.Int;
                    }

                case var case7 when case7 == typeof(long):
                case var case8 when case8 == typeof(ulong):
                    {
                        return AllowedDataTypes.Long;
                    }

                case var case9 when case9 == typeof(decimal):
                case var case10 when case10 == typeof(double):
                    {
                        return AllowedDataTypes.Decimal;
                    }

                case var case11 when case11 == typeof(string):
                case var case12 when case12 == typeof(char):
                    {
                        return AllowedDataTypes.String;
                    }

                case var case13 when case13 == typeof(byte[]):
                    {
                        return AllowedDataTypes.Blob;
                    }

                case var case14 when case14 == typeof(DateTime):
                    {
                        return AllowedDataTypes.DateTime;
                    }

                case var case15 when case15 == typeof(TimeSpan):
                    {
                        return AllowedDataTypes.TimeSpan;
                    }

                default:
                    {
                        return AllowedDataTypes.UNKNOWN;
                    }
            }
        }
    }
}